pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,238 @@
+		glog.Errorf("Failed to send registration status at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	if !resp.PluginRegistered {
+		glog.Errorf("Registration failed for plugin type: %s, name: %s, socket: %s, err: %s", infoResp.Type, infoResp.Name, socketPath, resp.Error)
+		return fmt.Errorf("%s", resp.Error)
+	}
+	glog.Infof("Successfully registered plugin for plugin type: %s, name: %s, socket: %s", infoResp.Type, infoResp.Name, socketPath)
+	return nil
+}
+
+func (w *Watcher) invokeRegistrationCallbackAtHandler(infoResp *registerapi.PluginInfo, socketPath string) *registerapi.RegistrationStatus {
+	resp := new(registerapi.RegistrationStatus)
+	if handlerCbkFn, ok := w.handlers[infoResp.Type]; ok {
+		var versions []string
+		for _, version := range infoResp.SupportedVersions {
+			versions = append(versions, version)
+		}
+		// calls handler callback to verify registration request
+		if verifyErr := handlerCbkFn(infoResp.Name, versions, socketPath); verifyErr != nil {
+			resp.PluginRegistered = false
+			resp.Error = fmt.Sprintf("Plugin registration failed with err: %v", verifyErr)
+		} else {
+			resp.PluginRegistered = true
+		}
+	} else {
+		resp.PluginRegistered = false
+		resp.Error = fmt.Sprintf("No handler found registered for plugin type: %s, socket: %s", infoResp.Type, socketPath)
+	}
+	return resp
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	w.stopCh = make(chan interface{})
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+
+	err = watcher.Add(w.path)
+	if err != nil {
+		watcher.Close()
+		return fmt.Errorf("failed to start plugin watcher, err: %v", err)
+	}
+	w.watcher = watcher
+	//startStatusCh := make(chan interface{})
+	w.wg.Add(1)
+	go func(watcher *fsnotify.Watcher) {
+		defer w.wg.Done()
+		//close(startStatusCh)
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Op&fsnotify.Create == fsnotify.Create {
+					w.registerPlugin(event.Name)
+				}
+				continue
+			case err := <-watcher.Errors:
+				//TODO: Handle errors by taking corrective measures
+				if err != nil {
+					glog.Errorf("Watcher received error: %v", err)
+				}
+				continue
+
+			case <-w.stopCh:
+				watcher.Close()
+				//close(w.stopCh)
+				break
+			}
+			break
+		}
+	}(watcher)
+	//<-startStatusCh
+	return nil
+}
+
+// Stop stops probing the creation of plugin sockets at the path
+func (w *Watcher) Stop() error {
+	//w.stopCh <- "done"
+	close(w.stopCh)
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		w.wg.Wait()
+	}()
+	select {
+	case <-c:
+	case <-time.After(5 * time.Second):
```

Comment:
The timeout is still disconnected from the underlying components timeout.
That timeout should be greater or equal to the underlying components (Dial) timeouts